grudge/op.py

```diff
@@ -48,37 +25,136 @@
 
 
 from numbers import Number
-from pytools import memoize_on_first_arg
+from functools import reduce
+
+from arraycontext import (
+        ArrayContext, FirstAxisIsElementsTag, make_loopy_program,
+        freeze)
+
+from grudge.discretization import DiscretizationCollection
 
-import numpy as np  # noqa
+from pytools import (
+    memoize_in,
+    memoize_on_first_arg,
+    keyed_memoize_in
+)
 from pytools.obj_array import obj_array_vectorize, make_obj_array
-import pyopencl.array as cla  # noqa
-from grudge import sym, bind
 
+from meshmode.dof_array import DOFArray
+
+import numpy as np
 import grudge.dof_desc as dof_desc
 
-from meshmode.mesh import BTAG_ALL, BTAG_NONE, BTAG_PARTITION  # noqa
-from meshmode.dof_array import freeze, flatten, unflatten
+from grudge.trace_pair import (  # noqa
+    interior_trace_pair,
+    interior_trace_pairs,
+    connected_ranks,
+    cross_rank_trace_pairs,
+    bdry_trace_pair,
+    bv_trace_pair
+)
+
+
+__doc__ = """
+Metadata and geometry
+^^^^^^^^^^^^^^^^^^^^^
+
+Kernel tags
+-----------
+
+.. autoclass:: HasElementwiseMatvecTag
```

Comment:
Not sure I'm ready to document this.

Reply:
https://github.com/inducer/grudge/pull/74/commits/d9bb98d696a00f727a4c29166f0debdfd1019b96